ThunkLibs/libwayland-client/Host.cpp

```diff
@@ -21,13 +22,281 @@ tags: thunklibs|wayland-client
+
+  memset((void*)host_interface.methods, 0, sizeof(wl_message) * host_interface.method_count);
+  for (int i = 0; i < host_interface.method_count; ++i) {
+    const auto& guest_method { from.data.methods.get_pointer()[i] };
+    host_layout<wl_message> host_method { guest_method };
+    fex_apply_custom_repacking_entry(host_method, guest_method);
+    memcpy((void*)&host_interface.methods[i], &host_method, sizeof(host_method));
+  }
+
+  memset((void*)host_interface.events, 0, sizeof(wl_message) * host_interface.event_count);
+  for (int i = 0; i < host_interface.event_count; ++i) {
+    const auto& guest_event { from.data.events.get_pointer()[i] };
+    host_layout<wl_message> host_event { guest_event };
+    fex_apply_custom_repacking_entry(host_event, guest_event);
+    memcpy((void*)&host_interface.events[i], &host_event, sizeof(host_event));
+  }
+}
+
+bool fex_custom_repack_exit(guest_layout<wl_interface>&, host_layout<wl_interface> const&) {
+  fprintf(stderr, "Should not be called: %s\n", __PRETTY_FUNCTION__);
+  std::abort();
+}
+void fex_custom_repack_entry(host_layout<wl_message>& into, guest_layout<wl_message> const& from) {
+  auto& host_method = into.data;
+  auto num_types = std::ranges::count_if(std::string_view { host_method.signature }, [](char c) { return std::isalpha(c); });
+  if (num_types) {
+    host_method.types = new const wl_interface*[num_types];
+    for (int type = 0; type < num_types; ++type) {
+      auto guest_interface_addr = from.data.types.get_pointer()[type];
+      host_method.types[type] = guest_interface_addr.force_get_host_pointer() ? lookup_wl_interface(guest_interface_addr) : nullptr;
+    }
+  }
+}
+bool fex_custom_repack_exit(guest_layout<wl_message>&, host_layout<wl_message> const&) {
+  fprintf(stderr, "Should not be called: %s\n", __PRETTY_FUNCTION__);
+  std::abort();
+}
+#else
+const wl_interface* lookup_wl_interface(guest_layout<const wl_interface*> interface) {
+  return interface.force_get_host_pointer();
+}
+#endif
+
+static wl_proxy* fexfn_impl_libwayland_client_wl_proxy_create(wl_proxy* proxy, guest_layout<const wl_interface*> guest_interface_raw) {
+  auto host_interface = lookup_wl_interface(guest_interface_raw);
+  return fexldr_ptr_libwayland_client_wl_proxy_create(proxy, host_interface);
+}
+
+#define WL_CLOSURE_MAX_ARGS 20
+static std::array<wl_argument, WL_CLOSURE_MAX_ARGS>
+fex_wl_remap_argument_list(guest_layout<wl_argument*> args) {
+  // TODO: Only copy as many arguments as needed by the message signature
+  std::array<wl_argument, WL_CLOSURE_MAX_ARGS> host_args;
+  for (int i = 0; i < host_args.size(); ++i) {
+    // NOTE: wl_argument can store a pointer argument, so for 32-bit guests
+    //       we need to make sure the upper 32-bits are explicitly zeroed
+    std::memset(&host_args[i], 0, sizeof(host_args[i]));
+    std::memcpy(&host_args[i], &args.get_pointer()[i], sizeof(args.get_pointer()[i]));
+  }
+  return host_args;
+}
```

Comment:
Should look into making this 32-bit-exclusive, since the extra copy isn't needed on 64-bit.